### ISSUE TITLE
Wss support

### ISSUE THREAD

**Issue by kingsleyh** [gbaranski/ezsockets#29]
Is WSS supported for client and server? And how would a simple example look?

**Comment by gbaranski** [comment]
Yes it is.
Use `wss://` prefix for the URL.
e.g 
```rust
let config = ClientConfig::new("wss://example.com");
let (handle, future) = ezsockets::connect(|_| Client {}, config).await;
```

**Comment by kingsleyh** [comment]
Thanks - I got it working but I had to add a dependency to my cargo.toml like this:

```
tokio-tungstenite = { version = "*", features = ["native-tls"] }
```

**Comment by gbaranski** [comment]
> Thanks - I got it working but I had to add a dependency to my cargo.toml like this:
> 
> ```
> tokio-tungstenite = { version = "*", features = ["native-tls"] }
> ```

Ah, right.
Now after 9867812f7d711f9f2f80be6cf03f5bde61dfcc8f you should be able to use `rustls` and `native-tls` feature.
It'll be released within v0.4.4.

**Comment by kingsleyh** [comment]
I'm trying to make a wss server example - but not quite sure how to do it - using TLS

e.g.

```
    let identity = Identity::from_pkcs12(include_bytes!("identity.pfx"), "password")?;
    let tls = TlsAcceptor::from(native_tls::TlsAcceptor::new(identity)?);
```

**Comment by kingsleyh** [comment]
full example: (not sure where to pass the tls)

```
use async_trait::async_trait;
use ezsockets::Error;
use ezsockets::Server;
use ezsockets::Socket;
use std::net::SocketAddr;
use async_native_tls::{Identity, TlsAcceptor};
use std::include_bytes;

type SessionID = u16;
type Session = ezsockets::Session<SessionID, ()>;

struct EchoServer {}

#[async_trait]
impl ezsockets::ServerExt for EchoServer {
    type Session = EchoSession;
    type Params = ();

    async fn accept(
        &mut self,
        socket: Socket,
        address: SocketAddr,
        _args: (),
    ) -> Result<Session, Error> {
        let id = address.port();
        let session = Session::create(|handle| EchoSession { id, handle }, id, socket);
        Ok(session)
    }

    async fn disconnected(
        &mut self,
        _id: <Self::Session as ezsockets::SessionExt>::ID,
    ) -> Result<(), Error> {
        Ok(())
    }

    async fn call(&mut self, params: Self::Params) -> Result<(), Error> {
        let () = params;
        Ok(())
    }
}

struct EchoSession {
    handle: Session,
    id: SessionID,
}

#[async_trait]
impl ezsockets::SessionExt for EchoSession {
    type ID = SessionID;
    type Args = ();
    type Params = ();

    fn id(&self) -> &Self::ID {
        &self.id
    }

    async fn text(&mut self, text: String) -> Result<(), Error> {
        self.handle.text(text);
        Ok(())
    }

    async fn binary(&mut self, _bytes: Vec<u8>) -> Result<(), Error> {
        unimplemented!()
    }

    async fn call(&mut self, params: Self::Params) -> Result<(), Error> {
        let () = params;
        Ok(())
    }
}



#[tokio::main]
async fn main() {
    let identity = Identity::from_pkcs12(include_bytes!("../ssc/certificate.pfx"), "password").unwrap();
    let tls = TlsAcceptor::from(native_tls::TlsAcceptor::new(identity).unwrap());
    
    tracing_subscriber::fmt::init();
    
    let (server, _) = Server::create(|_server| EchoServer {});
    // where to pass the tls? tungstenite docs say to pass it to tungstenite::accept
    ezsockets::tungstenite::run(server, "127.0.0.1:8080", |_| async move { Ok(()) })
        .await
        .unwrap();
}
```

**Comment by gbaranski** [comment]
> I'm trying to make a wss server example - but not quite sure how to do it - using TLS
> 
> e.g.
> 
> ```
>     let identity = Identity::from_pkcs12(include_bytes!("identity.pfx"), "password")?;
>     let tls = TlsAcceptor::from(native_tls::TlsAcceptor::new(identity)?);
> ```

Ah, you wanted a server with TLS. I'll be back in a moment with an example.

To be honest for TLS servers I'd suggest using some third-party reverse proxies like Caddy, Traefik or Nginx, as they can provide you automatic TLS Certificate generation.

**Comment by gbaranski** [comment]
TLS support with `native-tls` and `rustls` for servers will be available within v0.5.0 as #31 merges.
Now you can check the [existing example](https://github.com/gbaranski/ezsockets/tree/feat/tls-support/examples/echo-server-native-tls)

**Comment by kingsleyh** [comment]
ah amazing thanks - yeah you are right  - I will use Nginx to proxy the websocket - but great to see a direct example of tls - thanks very much :)
